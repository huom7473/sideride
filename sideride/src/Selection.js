import React from "react";
import { API, Auth } from "aws-amplify";
import { Alert, Button, Col, Container, Form, Row } from "react-bootstrap";
import AddressSearch from "./AddressSearch";
import qs from "qs";

export default class Selection extends React.Component {
    constructor(props) {
        super(props);
        if (this.props.location !== undefined) {
            console.log(this.props.location)
            let fromlat = qs.parse(this.props.location.search, { ignoreQueryPrefix: true }).fromlat;
            let fromlng = qs.parse(this.props.location.search, { ignoreQueryPrefix: true }).fromlng;
            let tolat = qs.parse(this.props.location.search, { ignoreQueryPrefix: true }).tolat;
            let tolng = qs.parse(this.props.location.search, { ignoreQueryPrefix: true }).tolng;
            this.state = {
                from: qs.parse(this.props.location.search, { ignoreQueryPrefix: true }).from,
                to: qs.parse(this.props.location.search, { ignoreQueryPrefix: true }).to,
                date: qs.parse(this.props.location.search, { ignoreQueryPrefix: true }).date,
                fromCoord: { lat: fromlat, lng: fromlng },
                toCoord: { lat: tolat, lng: tolng },
                showAlert: false,
                showDateAlert: false,
                info: {}
            };
            console.log(this.state.fromCoord)
        } else {
            this.state = { from: '', to: '', date: '', fromCoord: '', toCoord: '', showAlert: false, showDateAlert: false, info: {} };
        }
    }

    async componentDidMount() {
        const info = await Auth.currentUserInfo()
        this.setState({ info: info })
    }

    _handleUpdate = (evt) => {
        const { name, value } = evt.target;

        //this code is really nifty, it dynamically sets the input field with "name" to the corresponding value,
        //so it can update any of: email, password, confirmedPassword, errored...
        this.setState({ [name]: value }, () => {
            console.log("this.state", this.state);
        });
    };

    _handleFindRide = (evt) => {
        if (this.state.fromCoord === '' || this.state.toCoord === '' || this.state.date === '') {
            this.setState({ showAlert: true });
            return;
        }
        const initialChosen = new Date(this.state.date);
        const chosenDate = new Date(initialChosen.getTime() + initialChosen.getTimezoneOffset() * 60000);
        const today = new Date(new Date().toDateString());
        if (chosenDate < today) {
            this.setState({ showDateAlert: true });
            return;
        }
        else {
            this.props.history.push('/results?fromlat=' + this.state.fromCoord.lat + "&fromlng=" + this.state.fromCoord.lng + "&date=" + this.state.date
                + "&tolat=" + this.state.toCoord.lat + "&tolng=" + this.state.toCoord.lng + '&from=' + this.state.from + '&to=' + this.state.to);
            window.location.reload(false);
            //API.get('flaskapi ', '/api/find/' + this.state.info.username)
            //    .then((response) => console.log(response))
            evt.preventDefault();
        }
    };

    _handleCreateRide = (evt) => {
        this.props.history.push('/createride?from=' + this.state.from + "&to=" + this.state.to + "&date=" + this.state.date);

        //API Test
        API.get('flaskapi', '/api/createride?from=' + this.state.from + "&to=" + this.state.to + "&date=" + this.state.date)
            .then((response) => console.log(response))

        evt.preventDefault();
    };

    _handleToLatLong = ({ lat, lng, description }) => {
        this.setState({ toCoord: { lat, lng }, to: description });
    };

    _handleFromLatLong = ({ lat, lng, description }) => {
        this.setState({ fromCoord: { lat, lng }, from: description });
    };

    render() {
        if (!this.props.isLoaded) return "loading";
        else if (this.props.loadError) return "load error";
        return (
            <Container>
                <Alert className="floating-alert position-fixed"
                    show={this.state.showAlert}
                    variant="danger"
                    onClose={() => { this.setState({ showAlert: false }) }}
                    dismissible>
                    Please enter valid values for all required fields!
                </Alert>
                <Alert className="floating-alert position-fixed"
                    show={this.state.showDateAlert}
                    variant="danger"
                    onClose={() => { this.setState({ showDateAlert: false }) }}
                    dismissible>
                    Date entered was before today!
                </Alert>
                <Form>
                    <Form.Group as={Row} className="mb-3">
                        <Form.Label column sm={"auto"}>From:</Form.Label>
                        <Col>
                            <AddressSearch select={this._handleFromLatLong} initialValue={this.state.from} />
                        </Col>
                        <Form.Label column sm={"auto"}>To:</Form.Label>
                        <Col>
                            <AddressSearch select={this._handleToLatLong} initialValue={this.state.to} />
                        </Col>
                        <Form.Label column sm={"auto"}>Date:</Form.Label>
                        <Col>
                            <Form.Control name="date"
                                type="date"
                                value={this.state.date}
                                onChange={this._handleUpdate} />
                        </Col>
                    </Form.Group>
                </Form>
                <Button onClick={this._handleCreateRide}
                    className="me-2">Host a Ride</Button>
                <Button onClick={this._handleFindRide}>Find a Ride</Button>
            </Container>
        );
    }
}