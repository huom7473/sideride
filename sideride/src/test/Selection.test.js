import Selection from "../Selection"
import { configure, mount } from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import Amplify from 'aws-amplify';
import awsconfig from '../aws-exports';
import renderer from 'react-test-renderer';


Amplify.configure(awsconfig);
// Auth.signIn("test_user", "test_password");

configure({ adapter: new Adapter() });      // for enzyme methods like mount

const dummy_from = {
  lat: 12,
  lng: 34,
  description: '123 from street'
}

const dummy_to = {
  lat: 56,
  lng: 78,
  description: '123 to street'
}

const dummy_event = {
  target: { value: 'test-value' }
};

test('Selection: initialization', () => {
  const wrapper = mount(<Selection />);
  const instance = wrapper.instance();

  expect(instance.state.from).toEqual("");
  expect(instance.state.to).toEqual("");
  expect(instance.state.date).toEqual("");
});


test('Selection: _handleFromLatLong correctly sets from and fromCoord', () => {
  const wrapper = mount(<Selection />);
  const instance = wrapper.instance();

  instance._handleFromLatLong(dummy_from);
  expect(instance.state.from).toBe(dummy_from.description);
  expect(instance.state.fromCoord.lat).toBe(dummy_from.lat);
  expect(instance.state.fromCoord.lng).toBe(dummy_from.lng);
});

test('Selection: _handleToLatLong correctly sets to and toCoord', () => {
  const wrapper = mount(<Selection />);
  const instance = wrapper.instance();

  instance._handleToLatLong(dummy_to);
  expect(instance.state.to).toBe(dummy_to.description);
  expect(instance.state.toCoord.lat).toBe(dummy_to.lat);
  expect(instance.state.toCoord.lng).toBe(dummy_to.lng);
});

test('Selection: _handleFindRide correctly shows alerts for missing data or invalid dates', () => {
  const wrapper1 = mount(<Selection />);
  const instance1 = wrapper1.instance();
  instance1.state.fromCoord = '';
  instance1.state.toCoord = 'y';
  instance1.state.date = 'z';
  instance1._handleFindRide(dummy_event);
  expect(instance1.state.showAlert).toBe(true);
  expect(instance1.state.showDateAlert).toBe(false);

  const wrapper2 = mount(<Selection />);
  const instance2 = wrapper2.instance();
  instance2.state.fromCoord = 'x';
  instance2.state.toCoord = '';
  instance2.state.date = 'z';
  instance2._handleFindRide(dummy_event);
  expect(instance2.state.showAlert).toBe(true);
  expect(instance2.state.showDateAlert).toBe(false);

  const wrapper3 = mount(<Selection />);
  const instance3 = wrapper3.instance();
  instance3.state.fromCoord = 'x';
  instance3.state.toCoord = 'y';
  instance3.state.date = '';
  instance3._handleFindRide(dummy_event);
  expect(instance3.state.showAlert).toBe(true);
  expect(instance3.state.showDateAlert).toBe(false);

  const wrapper4 = mount(<Selection />);
  const instance4 = wrapper4.instance();
  instance4.state.fromCoord = 'x';
  instance4.state.toCoord = 'y';
  instance4.state.date = '2021-11-04';
  instance4._handleFindRide(dummy_event);
  expect(instance4.state.showAlert).toBe(false);
  expect(instance4.state.showDateAlert).toBe(true);

});

it('renders correctly selection component', () => {
  const SelectionComponent = renderer.create(<Selection />);
  const tree = SelectionComponent.toJSON();
  expect(tree).toMatchSnapshot();
});