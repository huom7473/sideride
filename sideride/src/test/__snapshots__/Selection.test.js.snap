// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly selection component 1`] = `"loading"`;
